### ISSUE TITLE
Users cannot add table metadata in the visual editor

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15283]
Users cannot add table metadata in the visual editor 

-   **Severity:** Medium
-   **Affected Populations:**
    -   Blind
    -   Low-Vision
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Editing
    -   Output Markup

#### Issue description

Authors using the visual editor can create a table, however the table
which results is only a series of rows and columns with no column-header
information, which may cause assistive technologies to assume these are
layout tables rather than data tables.

Users appear to be only able to add table headers by switching to the
HTML view. However not all authors have HTML knowledge and so may be
producing inaccessible tables.

Authors attempting to add table semantics such as a `<caption>`
element are shown an error, which may cause people using the editor to
think these are real errors.

##### Issue Code

```
    <!-- default resultant code in HTML view -->


    <table class="wp-block-table is-style-regular">


        <tbody>


            <tr>


                <td>This is a table!</td>


                <td>For tabular data</td>


            </tr>


            <tr>


                <td>But has no apparent way</td>


                <td>Of creating headers etc. without editing as HTML</td>


            </tr>


        </tbody>


    </table>
```

#### Remediation Guidance

Make the default Table block type assume that the first row is a header
row. This assumption is similar to what CSV files do, and will help
content authors create semantic tables despite any lack of HTML
knowledge.

Allow users to use the HTML editor to add in other semantics, such as
`<caption>` elements or `scope` attributes, without showing these
as errors to users.

##### Recommended Code

```
    <table class="wp-block-table is-style-regular">


        <thead>


            <tr>


                <th>This is a table!</th>


                <th>For tabular data</th>


            </tr>


        </thead>


        <tbody>


            <tr>


                <td>But has no apparent way</td>


                <td>Of creating headers etc. without editing as HTML</td>


            </tr>


        </tbody>


    </table>
```

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-103 in Tenon's report

**Comment by afercia** [comment]
This is a long standing issues and an accessibility priority to empower users to produce accessible content. See #6923 which incorporates #1470 (June 2017). Almost 2 years from the original report and still no method to define table headers.

At some point it was also suggested to discourage the usage of tabes, see https://github.com/WordPress/gutenberg/issues/1470#issuecomment-368303533

**Comment by talldan** [comment]
I've started working on an an issue (#15409) to add table header and footers to the table block. Needs some finessing, but I don't see any major blockers to getting it merged.

I haven't marked it as closing this issue as this also mentions adding captions to tables. I could potentially roll that into the PR, but I might just tackle it on a separate PR.

**Comment by talldan** [comment]
Here's a PR for the table caption: https://github.com/WordPress/gutenberg/pull/15554

**Comment by talldan** [comment]
After the caption PR is merged the main thing remaining (afaik) is the supporting the ability to add `scope` attributes to the table. 

The main thing at the moment would be to ensure those scope attributes aren't stripped off if a table is pasted in, or the html is manually edited.

At the moment the table implementation isn't really sophisticated enough to handle headers outside of a `<thead>` element when editing visually, so I don't think we need to worry about that aspect presently.

**Comment by mapk** [comment]
The PR, https://github.com/WordPress/gutenberg/pull/15554, looks ready for merging. I'll drop the label for design feedback. Excellent work!

**Comment by enriquesanchez** [comment]
@mapk Looks like #15554 was merged back in December. Can we close this issue now?

**Comment by mapk** [comment]
I believe so!

**Comment by afercia** [comment]
Thinking a follow-up issue should be created for further iterations. At the moment, the table block implementation is still pretty basic. Some important semantics can only by added by editing the HTML:

> At the moment the table implementation isn't really sophisticated enough to handle headers outside of a <thead> element when editing visually, so I don't think we need to worry about that aspect presently.

Also, the `caption` implementation seems seems incomplete to me, see https://github.com/WordPress/gutenberg/pull/15554#issuecomment-596426736

**Comment by joedolson** [comment]
I'm not totally convinced that this should have been closed, although the letter of the issue is resolved. Regardless, I agree that there should be a follow-up issue to iterate on this. It should be possible to define table row headers, not just a thead with column-scoped headers. 

 - It shouldn't be an option to leave <th> cells empty, or adding content should be prompted
 - Tables without any <th> cells should be marked as role="presentation"
 - The figure/figcaption implementation is a strange choice, given that there is a standard caption element (per @afercia's comment, indicated above.)

**Comment by mapk** [comment]
I think a follow-up issue would be great! Thank you 👍
